src/pypyjs.js

```diff
@@ -0,0 +1,1112 @@
+  let code_ptr;
+
+  return new Promise(function promise(resolve, reject) {
+    const _code = `try:
+  ${code.trim()}
+except Exception:
+  typ, val, tb = sys.exc_info()
+  err_name = getattr(typ, '__name__', str(typ))
+  err_msg = str(val)
+  err_trace = traceback.format_exception(typ, val, tb)
+  err_trace = err_trace[:1] + err_trace[2:]
+  err_trace = ''.join(err_trace)
+  js.globals['pypyjs']._lastErrorName = err_name
+  js.globals['pypyjs']._lastErrorMessage = err_msg
+  js.globals['pypyjs']._lastErrorTrace = err_trace
+`;
+
+    const code_chars = Module.intArrayFromString(_code);
+    code_ptr = Module.allocate(code_chars, 'i8', Module.ALLOC_NORMAL);
+    if (!code_ptr) {
+      throw new pypyjs.Error('Failed to allocate memory');
+    }
+
+    const res = Module._pypy_execute_source(code_ptr);
+    if (res < 0) {
+      throw new pypyjs.Error('Error executing python code');
+    }
+
+    Module._free(code_ptr);
+
+    // XXX TODO: races/re-entrancy on _lastError?
+    if (pypyjs._lastErrorName) {
+      const err = new pypyjs.Error(
+        pypyjs._lastErrorName,
+        pypyjs._lastErrorMessage,
+        pypyjs._lastErrorTrace
+      );
+      pypyjs._lastErrorName = null;
+      pypyjs._lastErrorMessage = null;
+      pypyjs._lastErrorTrace = null;
+      reject(err);
+    }
+
+    resolve(null);
+  });
+};
+
+// Method to determine when the interpreter is ready.
+//
+// This method returns a promise that will resolve once the interpreter
+// is ready for use.
+//
+pypyjs.prototype.ready = function ready() {
+  return this._ready;
+};
+
+// Method to execute some python code.
+//
+// This passes the given python code to the VM for execution.
+// It's fairly directly analogous to the 'exec" statement in python.
+// It is not possible to directly access the result of the code, if any.
+// Rather you should store it into a variable and then use the get() method.
+//
+pypyjs.prototype.exec = function exec(code, options) {
+  return this._ready.then(() => {
+    let promise = Promise.resolve();
+    let preCode;
+
+    // Find any "import" statements in the code,
+    // and ensure the modules are ready for loading.
+    if (this.autoLoadModules) {
+      promise = promise.then(() => {
+        return this.findImportedNames(code);
+      })
+      .then((imports) => {
+        return this.loadModuleData.apply(this, imports);
+      });
+    }
+
+    // if any modules have been re-added then we need to remove them from
+    // sys.modules which clears them from memory and allows them to be reloaded
+    // from the emscripten file system
+    if (Object.keys(this._modulesToReset).length) {
+      // construct python code to remove module from sys.modules:
+      // ```python
+      //   import sys
+      //   if 'foo' in sys.modules: del(sys.modules['foo'])
+      // ```
+      const modulesToLoad =
+        Object.keys(this._modulesToReset)
+          .map(mod => `if '${mod}' in sys.modules: del(sys.modules['${mod}'])`);
```

Comment:
It might also work as a `resetModule` method that the calling code can use when it knows it needs to reset the state, if this is an operation that needs to be done frequently.

In general though, importing modules can have side-effects that are not undone by `del sys.modules`, so re-creating a fresh VM is likely to be more reliable.